src/rpc/client.rs

```diff
@@ -0,0 +1,264 @@
+        let crate::rpc_client::RpcRequest {
+            method_name,
+            params,
+            api_version,
+            timeout,
+            ..
+        } = req;
+
+        let client = self.get_or_init_client(api_version).await?;
+        let span = tracing::debug_span!("request", method = %method_name, url = %client.url);
+        let work = async {
+            // jsonrpsee's clients have a global `timeout`, but not a per-request timeout, which
+            // RpcRequest expects.
+            // So shim in our own timeout
+            let result_or_timeout = tokio::time::timeout(
+                timeout,
+                match params {
+                    serde_json::Value::Null => {
+                        client.request::<T::LotusJson, _>(method_name, ArrayParams::new())
+                    }
+                    serde_json::Value::Array(it) => {
+                        let mut params = ArrayParams::new();
+                        for param in it {
+                            params.insert(param)?
+                        }
+                        trace_params(params.clone());
+                        client.request(method_name, params)
+                    }
+                    serde_json::Value::Object(it) => {
+                        let mut params = ObjectParams::new();
+                        for (name, param) in it {
+                            params.insert(&name, param)?
+                        }
+                        trace_params(params.clone());
+                        client.request(method_name, params)
+                    }
+                    prim @ (serde_json::Value::Bool(_)
+                    | serde_json::Value::Number(_)
+                    | serde_json::Value::String(_)) => {
+                        return Err(ClientError::Custom(format!(
+                            "invalid parameter type: `{}`",
+                            prim
+                        )))
+                    }
+                },
+            )
+            .await;
+            let result = match result_or_timeout {
+                Ok(Ok(it)) => Ok(T::from_lotus_json(it)),
+                Ok(Err(e)) => Err(e),
+                Err(_) => Err(ClientError::RequestTimeout),
+            };
+            debug!(?result);
+            result
+        };
+        work.instrument(span.or_current()).await
+    }
+    async fn get_or_init_client(&self, version: ApiVersion) -> Result<&OneClient, ClientError> {
+        match version {
+            ApiVersion::V0 => &self.v0,
+            ApiVersion::V1 => &self.v1,
+        }
+        .get_or_try_init(|| async {
+            let url = self
+                .base_url
+                .join(match version {
+                    ApiVersion::V0 => "rpc/v0",
+                    ApiVersion::V1 => "rpc/v1",
+                })
+                .map_err(|it| {
+                    ClientError::Custom(format!("creating url for endpoint failed: {}", it))
```

Comment:
Either that or let's keep it like here in `impl OneClient` as well for consistency.
```suggestion
                    ClientError::Custom(format!("creating url for endpoint failed: {it}"))
```